### PR TITLE
WebP/Lossy: Add code to read the frame header in the first partition

### PR DESCRIPTION
**Pull request by nico**

### CODE REVIEW

**Comment by nico** [comment]
For reading this, I'd recommend opening https://datatracker.ietf.org/doc/html/rfc6386#section-19.2 and looking at decode_VP8_frame_header() and comparing the two.

Hm, maybe I should reorder the helpers to match the order of the helpers in the spec. Let me do that.

**Review by skyrising — APPROVED** [review]

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp [hunk 1]
```diff
@@ -99,16 +101,312 @@ ErrorOr<VP8Header> decode_webp_chunk_VP8_header(ReadonlyBytes vp8_data)
     return VP8Header { version, show_frame, size_of_first_partition, width, horizontal_scale, height, vertical_scale, vp8_data.slice(10) };
 }
 
+namespace {
+
+// Reads n bits followed by a sign bit (0: positive, 1: negative).
+ErrorOr<i8> read_signed_literal(BooleanDecoder& decoder, u8 n)
```
Comment (ADKaster) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:107]: Does this method need help to ensure we don't truncate a value that's > 7 bits?

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp [hunk 2]
```diff
@@ -99,16 +101,312 @@ ErrorOr<VP8Header> decode_webp_chunk_VP8_header(ReadonlyBytes vp8_data)
...
+    dbgln_if(WEBP_DEBUG, "log2_nbr_of_dct_partitions {}", log2_nbr_of_dct_partitions);
+    header.number_of_dct_partitions = 1 << log2_nbr_of_dct_partitions;
+
+    header.quantization_indices = TRY(decode_VP8_frame_header_quantization_indices(decoder));
+
+    // In the VP8 spec, this is in an `if (key_frames)` followed by a lengthy `else`, but webp files only have key frames.
+    u8 refresh_entropy_probs = TRY(L(1)); // Has no effect in webp files.
+    dbgln_if(WEBP_DEBUG, "refresh_entropy_probs {}", refresh_entropy_probs);
+
+    memcpy(header.coefficient_probabilities, default_coeff_probs, sizeof(header.coefficient_probabilities));
```
Comment (ADKaster) [inline Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:237]: Any reason the compiler-generated copy ctor for this POD struct isn't sufficient? I think elsewhere we've avoided memcpy for that usage. Or that we couldn't `= default_coeff_probs` in the designated initializer for the FrameHeader struct.

**Review by ADKaster — COMMENTED** [review]
2 style questions, but overall looks nice 👍

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp (hunk 1 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:107]: Added a `VERIFY(n <= 7)` (the 5 callers this function currently has are all there ever will be; all 5 hardcode a fixed n value <= 7.)

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp (hunk 2 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:238]: it don't build:

```
WebPLoaderLossy.cpp:239:38: error: array type 'Gfx::(anonymous namespace)::CoefficientProbabilities' (aka 'unsigned char[4][8][3][11]') is not assignable
    header.coefficient_probabilities = default_coeff_probs;
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
```

(repro:

```sh
% cat test.cc
using t = int[2][2];

struct s {
    t k;
};

int main() {
    int a[2][2] = { { 1, 2}, {3, 4} };

    s k;
    k.k = a;
}
% clang -c test.cc -std=c++11
test.cc:11:9: error: array type 't' (aka 'int[2][2]') is not assignable
    k.k = a;
    ~~~ ^
```

**Review by nico — COMMENTED** [review]
ty!

File: Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp (hunk 2 above)
Comment (nico) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:238]: (ie it ain't a struct but an array)

Comment (ADKaster) [reply Userland/Libraries/LibGfx/ImageFormats/WebPLoaderLossy.cpp:238]: ah, oki. Well, we could "prefer AK::Array to C-style arrays" but I don't think that one is actually written down (and I don't care *that* much myself).

**Review by ADKaster — APPROVED** [review]

**Pull request merged**